### PR TITLE
Feat/redeem payment command 2850

### LINKED ISSUE #2850: Create `redeem-payment` command.
### Description

We currently have no reasonable way for a miner to redeem the payment vouchers contained within a deal. When a miner invokes:

``` bash
go-filecoin deals redeem <deal id>
```

The command will

1. Find the deal
2. Return error if no deal is found or deal contains no vouchers
3. Find the first voucher with the greatest value whose validAt is not less than the current block height
4. Return if no voucher is found
5. Return if the vouchers value is not greater than its payment channel's `AmmountRedeemed`
6. Send a `paych redeem` message with the voucher. If a `verifyPieceInclusion` condition is present, it will also submit the sector id and piece inclusion proof. 
7. Returns the CID of the submitted message.

### Acceptance criteria

1. The command exists and works.
2. Most of the command is implemented in porcelain.
3. The porcelain is well tested
4. The command has daemon tests written in FAST

### PR DESCRIPTION
**Pull request by rosalinekarr**
# Problem

We don't have a simple way for users to redeem vouchers for a given storage deal.

# Solution

Create a `go-filecoin deals redeem` command for redeeming the first valid voucher on a given deal.

Resolves #2850

### CODE REVIEW

**Comment by acruikshank** [comment]
@rosalinekarr there are changes in this PR that aren't strictly necessary for #2850. Please describe what they are and why they are in this PR.

**Comment by rosalinekarr** [comment]
@acruikshank The only other thing I addressed in this PR was moving the `go-filecoin client list-deals` and `go-filecoin miner list-deals` commands into a new `go-filecoin deals list` command. The functionality and tests for these commands are unchanged, just moved. I included these changes because they seemed small and related enough to creating the `go-filecoin deals` command to justify making them.

If you want to ignore these changes, just skip the first two commits to the branch when diffing, or if you'd like, I can remove them and make a new PR for these changes.

**Comment by rosalinekarr** [comment]
@acruikshank never mind, I went ahead and extracted it to a new PR here: https://github.com/filecoin-project/go-filecoin/pull/2910

File: porcelain/storagedeals.go [hunk 1]
```diff
@@ -140,3 +142,84 @@ func filterDealChannel(dealCh <-chan *StorageDealLsResult, filterFunc func(*stor
...
+type dealRedeemPlumbing interface {
+	ChainBlockHeight() (*types.BlockHeight, error)
+	DealGet(context.Context, cid.Cid) (*storagedeal.Deal, error)
+	MessagePreview(context.Context, address.Address, address.Address, string, ...interface{}) (types.GasUnits, error)
+	MessageSendWithDefaultAddress(context.Context, address.Address, address.Address, *types.AttoFIL, types.AttoFIL, types.GasUnits, string, ...interface{}) (cid.Cid, error)
+}
+
+// DealRedeem redeems a voucher for the deal with the given cid and returns
+// either the cid of the created redeem message or an error
+func DealRedeem(ctx context.Context, plumbing dealRedeemPlumbing, fromAddr address.Address, dealCid cid.Cid, gasPrice types.AttoFIL, gasLimit types.GasUnits) (cid.Cid, error) {
```
Comment (acruikshank) [inline porcelain/storagedeals.go]: This plumbing function needs a test to confirm it correctly chooses the best voucher from a deal and sends a properly formatted message.

File: porcelain/storagedeals.go [hunk 2]
```diff
@@ -140,3 +142,84 @@ func filterDealChannel(dealCh <-chan *StorageDealLsResult, filterFunc func(*stor
...
+		gasPrice,
+		gasLimit,
+		"redeem",
+		params...,
+	)
+}
+
+// DealRedeemPreview previews the redeem method for a deal and returns the
+// expected gas used
+func DealRedeemPreview(ctx context.Context, plumbing dealRedeemPlumbing, fromAddr address.Address, dealCid cid.Cid) (types.GasUnits, error) {
```
Comment (acruikshank) [inline porcelain/storagedeals.go]: This plumbing function needs a test to confirm that it creates the deal redeem params the same way as `DealRedeem` and properly formats a message sent to `MessagePreview`.

File: commands/deals_daemon_test.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	redeemCid, err := minerDaemon.DealsRedeem(ctx, proposeDealOutput.ProposalCid, fast.AOPrice(big.NewFloat(0.001)), fast.AOLimit(100))
+	require.NoError(t, err)
+
+	_, err = clientDaemon.MessageWait(ctx, redeemCid)
+	require.NoError(t, err)
+
+	newWalletBalance, err := minerDaemon.WalletBalance(ctx, minerAddress)
+	require.NoError(t, err)
+
+	expectedBalanceDiff := types.NewAttoFILFromFIL(0).String()
```
Comment (acruikshank) [inline commands/deals_daemon_test.go]: This shouldn't be zero. It should be the value of the voucher.

File: porcelain/storagedeals_test.go [hunk 4]
```diff
@@ -139,3 +140,162 @@ func TestDealMinerLs(t *testing.T) {
...
+		expectedVoucher: expectedVoucher,
+		fromAddr:        fromAddr,
+		messageCid:      messageCid,
+		vouchers:        vouchers,
+	}
+
+	resultCid, err := porcelain.DealRedeem(context.Background(), plumbing, fromAddr, dealCid, types.NewAttoFILFromFIL(0), types.NewGasUnits(0))
+	require.NoError(t, err)
+
+	assert.Equal(t, messageCid, resultCid)
```
Comment (acruikshank) [inline porcelain/storagedeals_test.go]: Running the equality assertions in the stub obscures the meaning of this test. It's correct and sufficient, but it takes some unnecessary analysis on the readers part to establish what's under test.

**Pull request merged**